#  EPA-122A Introduction to *Spatial* Data Science


## Assignment 1: Data Collection and Wrangling


---



# ``Instructions``

This assignment puts together what you learned in **Weeks 1-2**. You will be working with a dataset which is in the form of a spreadsheet. It may contain many different data types in the columns. All data frames contain column names, which are strings, and row indices, which are integers. In this assignment you will illustrate your knwoledge about bundling various kinds of data together to be able to do higher-level tasks.

_Note:_ Go through **labs and homeworks 00-02** before starting this assignment.

#### 1.1 Submission

Please submit the results by Brightspace under **Assignment 01**, using a single file as example,

```text
firstname_secondname_thirdname_lastname_01.html

```

**If your file is not named in lowercase letters as mentioned above, your assignment will not be read by the script that works to compile > 200 assignments and you will miss out on the grades. I don't want that, so be exceptionally careful that you name it properly. Don't worry if you spelled your name incorrectly. I want to avoid a situation where I have 200 assignments all called assignment_01.html**

Please **do not** submit any data or files other than the ``html file``.

#### 1.2 How do you convert to HTML?

There are 2 ways,

1. from a running notebook, you can convert it into html by clicking on the file tab on the main menu of Jupyter Lab
    * File &rightarrow; Export Notebooks as... &rightarrow; Export Notebook to HTML
2. go to terminal or command line and type
    * ``jupyter nbconvert --to html <notebook_name>.ipynb  ``


#### 1.3 Learning Objectives

This assignment is designed to support three different learning objectives. After completing the following exercises you will be able to:

* Explore variables in a dataset
* Manage missing data
* Reshape data to get it in a form useful for statistical analysis

#### 1.4 Tasks

This assignment requires you to go through five tasks in cleaning your data.

1. Reading and summarizing the data.
2. Subsetting the Data. This extracts just the part of the data you want to analyse.
3. Manage Missing Data. Some data is not available for all objects of interest (rows) or all variables for every object (columns).
4. Shape the Data. We need to convert the data into a suitable format for analysis.
5. Saving the Results. The results are saved for future use.

<br/>

***

# ``Critical Data Science``

Throughout the assignment, we encourage you to critically reflect on your choices during the Data Science process. To help you set-up your Critical Data Science process, we have provided you with a 'Guide on Critical Data Science'. Section 3.2 contains a step-by-step approach which each key considerations for each part of the data science process. The guide can be found here: https://epa122a.github.io/resources_index.html#a-guide-to-critical-data-science. Throughout this exercises, you will also find several questions which you can use to reflect on your data science choices.

<br/>

***

# ``Task 1: Downloading the Data``

For this assignment we are going to use the Den Haag cijfers database as a source of data. The database allows us to investigate a range of socio-economic variables for residents in The Hague. The data consists of time series which in some cases dates back into the seventies.

You can download the data here as a csv file (You will have to choose the variables and file format yourself):
https://denhaag.incijfers.nl/jive

Put the data in a convenient location on your computer or laptop, ideally in a folder called **data** which is next to this **jupyter notebook**. I recommend taking a look at the file in a text editor like _zed_ for any system or notepad++ for windows. These will also make your life easy for everything else on your computer. It’s a big file and it may take a while to load onto your laptop and into Python (running on the jupyter labs environment).

## ``Exercise: Downloading the Data``

**IMPORTANT** make sure your code can run independent of the machine. i.e.
- Use relative path links instead of absolute paths. If your data folder is named C:/HelloKitty/MyGummyBears/IlovePython/DenHaagData.csv, then your program will not be reproducible on any other machine. Check out this very easy to follow and handy guide on [relative paths](https://www.delftstack.com/howto/python/relative-path-in-python/).
- Organise the data in a folder called `data` and run your notebook next to it organised as follows

```text
├── trivik_verma_01.ipynb
├── data
│   ├── DenHaagData.csv
```

- Load the csv file into Python
- Explore it by looking at first and last 5 rows
- Programatically find and print information on the data,
    - number of columns in the data
    - names of the columns in the data
    - number of rows in the data (excluding the header names)
    - how many unique neighborhoods in the data
    - how many unique variables are in the data

In [1]:
import pandas as pd

# Step 1: Load the CSV file
data = pd.read_csv('data/DenHaagData.csv', na_values=['-', 'x'])
       
data.head(5)

District_name  Year  \
0  14 Bohemen, Meer en Bos  2014   
1  14 Bohemen, Meer en Bos  2015   
2  14 Bohemen, Meer en Bos  2016   
3  14 Bohemen, Meer en Bos  2017   
4  14 Bohemen, Meer en Bos  2018   

   Gemiddelde totaalscore Sociaal Economische Status WOA (SES/WOA score) [score]  \
0                                              0.127                               
1                                              0.125                               
2                                              0.121                               
3                                              0.119                               
4                                              0.110                               

   Gemiddelde SES/WOAscore: deelscore financiële welvaart [score]  \
0                                              0.105                
1                                              0.104                
2                                              0.092                
3                                              0.087                
4                                              0.085                

   Gemiddelde SES/WOAscore: deelscore opleidingsniveau [score]  \
0                                              0.034             
1                                              0.035             
2                                              0.037             
3                                              0.041             
4                                              0.041             

   Gemiddelde SES/WOAscore: deelscore arbeidsverleden [score]  \
0                                             -0.013            
1                                             -0.013            
2                                             -0.007            
3                                             -0.009            
4                                             -0.016            

   Gross population density [residents/hectare]  0 to 19-year-olds [persons]  \
0                                          38.5                        627.0   
1                                          39.1                        602.0   
2                                          38.9                        612.0   
3                                          39.3                        622.0   
4                                          40.0                        648.0   

   20 to 64-year-olds [persons]  65 year-olds and above [persons]  ...  \
0                        2408.0                            1637.0  ...   
1                        2399.0                            1743.0  ...   
2                        2373.0                            1773.0  ...   
3                        2400.0                            1793.0  ...   
4                        2415.0                            1829.0  ...   

   Overlast ivm alcohol/drugs [number]  Openbare dronkenschap [number]  \
0                                  0.0                             1.0   
1                                  2.0                             1.0   
2                                  5.0                             0.0   
3                                 10.0                             2.0   
4                                 12.0                             2.0   

   Overlast zwervers [number]  Overlast door verward persoon [number]  \
0                         0.0                                    17.0   
1                         0.0                                    26.0   
2                         4.0                                    32.0   
3                        10.0                                    62.0   
4                        10.0                                    47.0   

   Geluidsoverlast [number]  Rapportcijfer 'Veiligheid in de buurt' [score]  \
0                       9.0                                             NaN   
1                       7.0                                             NaN   
2                       6.0              

In [2]:
# Full translation mapping for all 77 entries

column_translation = {
        "District_name": "district_name",
        "Year": "Year",
        "Gemiddelde totaalscore Sociaal Economische Status WOA (SES/WOA score) [score]": "Average Total Score Socioeconomic Status WOA (SES/WOA Score) [score]",
        "Gemiddelde SES/WOAscore: deelscore financiële welvaart [score]": "Average SES/WOA Score: Financial Prosperity Subscore [score]",
        "Gemiddelde SES/WOAscore: deelscore opleidingsniveau [score]": "Average SES/WOA Score: Educational Level Subscore [score]",
        "Gemiddelde SES/WOAscore: deelscore arbeidsverleden [score]": "Average SES/WOA Score: Work History Subscore [score]",
        "Gross population density [residents/hectare]": "Gross Population Density [residents/hectare]",
        "0 to 19-year-olds [persons]": "0 to 19-Year-Olds [persons]",
        "20 to 64-year-olds [persons]": "20 to 64-Year-Olds [persons]",
        "65 year-olds and above [persons]": "65-Year-Olds and Above [persons]",
        "% 0 to 19-year-olds [procent (%)]": "Percentage of 0 to 19-Year-Olds [percent (%)]",
        "% 20 to 64-year-olds [procent (%)]": "Percentage of 20 to 64-Year-Olds [percent (%)]",
        "% 65-year-olds and above [procent (%)]": "Percentage of 65-Year-Olds and Above [percent (%)]",
        "Average age of population [age]": "Average Age of Population [age]",
        "Dekkingsgraad [procent (%)]": "Coverage Ratio [percent (%)]",
        "Opleidingsniveau Laag [procent (%)]": "Educational Level Low [percent (%)]",
        "Opleidingsniveau Middelbaar [procent (%)]": "Educational Level Intermediate [percent (%)]",
        "Opleidingsniveau Hoog [procent (%)]": "Educational Level High [percent (%)]",
        "Average disposable income of private households [euro_s]": "Average Disposable Income of Private Households [euros]",
        "Average standardised disposable income of private households [euro_s]": "Average Standardised Disposable Income of Private Households [euros]",
        "Average personal income of individuals, 52 weeks [euro_s]": "Average Personal Income of Individuals, 52 Weeks [euros]",
        "Geregistreerde werkzoekenden UWV zonder een dienstverband [number]": "Registered Job Seekers UWV Without Employment [number]",
        "Geregistreerde werkzoekenden UWV met een dienstverband [number]": "Registered Job Seekers UWV With Employment [number]",
        "aantal 0 tot 23 jarigen in Haagse bevolking [persons]": "Number of 0 to 23-Year-Olds in The Hague Population [persons]",
        "% 0 tot 23 jarigen in Haagse bevolking [procent (%)]": "Percentage of 0 to 23-Year-Olds in The Hague Population [percent (%)]",
        "aantal 0 tot 4 jarigen in Haagse bevolking [persons]": "Number of 0 to 4-Year-Olds in The Hague Population [persons]",
        "aantal 4 tot 12 jarigen in Haagse bevolking [persons]": "Number of 4 to 12-Year-Olds in The Hague Population [persons]",
        "aantal 12 tot 18 jarigen in Haagse bevolking [persons]": "Number of 12 to 18-Year-Olds in The Hague Population [persons]",
        "aantal 18 tot 23 jarigen in Haagse bevolking [persons]": "Number of 18 to 23-Year-Olds in The Hague Population [persons]",
        "Afstand tot basisschool in km [kilometres]": "Distance to Primary School [kilometres]",
        "Distance to child day care centre [kilometres]": "Distance to Child Day Care Centre [kilometres]",
        "Distance to out-of-school care [kilometres]": "Distance to Out-of-School Care [kilometres]",
        "Distance to secondary school, total [kilometres]": "Distance to Secondary School, Total [kilometres]",
        "Distance to VMBO school [kilometres]": "Distance to VMBO School [kilometres]",
        "Distance to HAVO/VWO school [kilometres]": "Distance to HAVO/VWO School [kilometres]",
        "Number of child day care centres within 1 km [number]": "Number of Child Day Care Centres Within 1 km [number]",
        "Number of child day care centres within 3 km [number]": "Number of Child Day Care Centres Within 3 km [number]",
        "Number of child day care centres within 5 km [number]": "Number of Child Day Care Centres Within 5 km [number]",
        "Number of out-of-school care centres within 1 km [number]": "Number of Out-of-School Care Centres Within 1 km [number]",
        "Number of out-of-school care centres within 3 km [number]": "Number of Out-of-School Care Centres Within 3 km [number]",
        "Number of out-of-school care centres within 5 km [number]": "Number of Out-of-School Care Centres Within 5 km [number]",
        "Number of schools within 1 km [number]": "Number of Schools Within 1 km [number]",
        "Number of schools within 3 km [number]": "Number of Schools Within 3 km [number]",
        "Number of schools within 5 km [number]": "Number of Schools Within 5 km [number]",
        "Total number of secondary schools within 3 km [number]": "Total Number of Secondary Schools Within 3 km [number]",
        "Total number of secondary schools within 5 km [number]": "Total Number of Secondary Schools Within 5 km [number]",
        "Total number of secondary schools within 10 km [number]": "Total Number of Secondary Schools Within 10 km [number]",
        "Number of VMBO schools within 3 km [number]": "Number of VMBO Schools Within 3 km [number]",
        "Number of VMBO schools within 5 km [number]": "Number of VMBO Schools Within 5 km [number]",
        "Number of VMBO schools within 10 km [number]": "Number of VMBO Schools Within 10 km [number]",
        "Number of HAVO/VWO schools within 3 km [number]": "Number of HAVO/VWO Schools Within 3 km [number]",
        "Number of HAVO/VWO schools within 5 km [number]": "Number of HAVO/VWO Schools Within 5 km [number]",
        "Number of HAVO/VWO schools within 10 km [number]": "Number of HAVO/VWO Schools Within 10 km [number]",
        "Number of participants in education [persons]": "Number of Participants in Education [persons]",
        "Number of new school drop-outs [persons]": "Number of New School Drop-Outs [persons]",
        "% (nieuwe) voortijdig schoolverlaters [procent (%)]": "Percentage of New School Drop-Outs [percent (%)]",
        "Number of participants in secondary education [persons]": "Number of Participants in Secondary Education [persons]",
        "Number of new drop-outs from secondary education [persons]": "Number of New Drop-Outs from Secondary Education [persons]",
        "% new school drop-outs, secondary education [procent (%)]": "Percentage of New School Drop-Outs, Secondary Education [percent (%)]",
        "Number of participants in MBO [persons]": "Number of Participants in MBO [persons]",
        "Number of new drop-outs from MBO [persons]": "Number of New Drop-Outs from MBO [persons]",
        "% new school drop-outs, MBO [procent (%)]": "Percentage of New School Drop-Outs, MBO [percent (%)]",
        "aantal onderwijsdeelnemers VAVO [persons]": "Number of Education Participants VAVO [persons]",
        "aantal nieuwe voortijdig schoolverlaters VAVO [persons]": "Number of New Drop-Outs from VAVO [persons]",
        "% nieuwe voortijdig schoolverlaters VAVO [procent (%)]": "Percentage of New Drop-Outs from VAVO [percent (%)]",
        "Overlast totaal [number]": "Total Nuisance [number]",
        "Jeugdoverlast [number]": "Youth Nuisance [number]",
        "Overlast ivm alcohol/drugs [number]": "Nuisance Related to Alcohol/Drugs [number]",
        "Openbare dronkenschap [number]": "Public Drunkenness [number]",
        "Overlast zwervers [number]": "Nuisance from Homeless People [number]",
        "Overlast door verward persoon [number]": "Nuisance from Disturbed Persons [number]",
        "Geluidsoverlast [number]": "Noise Nuisance [number]",
        "Rapportcijfer 'Veiligheid in de buurt' [score]": "Score for 'Safety in the Neighborhood' [score]",
        "Voelt zich wel eens onveilig in eigen buurt [procent (%)]": "Feels Unsafe in Own Neighborhood Occasionally [percent (%)]",
        "Voelt zich vaak onveilig in eigen buurt [procent (%)]": "Feels Unsafe in Own Neighborhood Frequently [percent (%)]",
        "Voelt zich wel eens onveilig in algemeen [procent (%)]": "Feels Unsafe in General Occasionally [percent (%)]",
        "Voelt zich vaak onveilig in algemeen [procent (%)]": "Feels Unsafe in General Frequently [percent (%)]"
    }


data.rename(columns=column_translation, inplace=True)



# Define the data types for each column

column_data_types = {
        "district_name": str,
        "Year": "Int64",  # Nullable integer
        "Average Total Score Socioeconomic Status WOA (SES/WOA Score) [score]": float,
        "Average SES/WOA Score: Financial Prosperity Subscore [score]": float,
        "Average SES/WOA Score: Educational Level Subscore [score]": float,
        "Average SES/WOA Score: Work History Subscore [score]": float,
        "Gross Population Density [residents/hectare]": float,
        "0 to 19-Year-Olds [persons]": "Int64",
        "20 to 64-Year-Olds [persons]": "Int64",
        "65-Year-Olds and Above [persons]": "Int64",
        "Percentage of 0 to 19-Year-Olds [percent (%)]": float,
        "Percentage of 20 to 64-Year-Olds [percent (%)]": float,
        "Percentage of 65-Year-Olds and Above [percent (%)]": float,
        "Average Age of Population [age]": float,
        "Coverage Ratio [percent (%)]": float,
        "Educational Level Low [percent (%)]": float,
        "Educational Level Intermediate [percent (%)]": float,
        "Educational Level High [percent (%)]": float,
        "Average Disposable Income of Private Households [euros]": float,
        "Average Standardised Disposable Income of Private Households [euros]": float,
        "Average Personal Income of Individuals, 52 Weeks [euros]": float,
        "Registered Job Seekers UWV Without Employment [number]": "Int64",
        "Registered Job Seekers UWV With Employment [number]": "Int64",
        "Number of 0 to 23-Year-Olds in The Hague Population [persons]": "Int64",
        "Percentage of 0 to 23-Year-Olds in The Hague Population [percent (%)]": float,
        "Number of 0 to 4-Year-Olds in The Hague Population [persons]": "Int64",
        "Number of 4 to 12-Year-Olds in The Hague Population [persons]": "Int64",
        "Number of 12 to 18-Year-Olds in The Hague Population [persons]": "Int64",
        "Number of 18 to 23-Year-Olds in The Hague Population [persons]": "Int64",
        "Distance to Primary School [kilometres]": float,
        "Distance to Child Day Care Centre [kilometres]": float,
        "Distance to Out-of-School Care [kilometres]": float,
        "Distance to Secondary School, Total [kilometres]": float,
        "Distance to VMBO School [kilometres]": float,
        "Distance to HAVO/VWO School [kilometres]": float,
        "Number of Child Day Care Centres Within 1 km [number]": "Int64",
        "Number of Child Day Care Centres Within 3 km [number]": "Int64",
        "Number of Child Day Care Centres Within 5 km [number]": "Int64",
        "Number of Out-of-School Care Centres Within 1 km [number]": "Int64",
        "Number of Out-of-School Care Centres Within 3 km [number]": "Int64",
        "Number of Out-of-School Care Centres Within 5 km [number]": "Int64",
        "Number of Schools Within 1 km [number]": "Int64",
        "Number of Schools Within 3 km [number]": "Int64",
        "Number of Schools Within 5 km [number]": "Int64",
        "Total Number of Secondary Schools Within 3 km [number]": "Int64",
        "Total Number of Secondary Schools Within 5 km [number]": "Int64",
        "Total Number of Secondary Schools Within 10 km [number]": "Int64",
        "Number of VMBO Schools Within 3 km [number]": "Int64",
        "Number of VMBO Schools Within 5 km [number]": "Int64",
        "Number of VMBO Schools Within 10 km [number]": "Int64",
        "Number of HAVO/VWO Schools Within 3 km [number]": "Int64",
        "Number of HAVO/VWO Schools Within 5 km [number]": "Int64",
        "Number of HAVO/VWO Schools Within 10 km [number]": "Int64",
        "Number of Participants in Education [persons]": "Int64",
        "Number of New School Drop-Outs [persons]": "Int64",
        "Percentage of New School Drop-Outs [percent (%)]": float,
        "Number of Participants in Secondary Education [persons]": "Int64",
        "Number of New Drop-Outs from Secondary Education [persons]": "Int64",
        "Percentage of New School Drop-Outs, Secondary Education [percent (%)]": float,
        "Number of Participants in MBO [persons]": "Int64",
        "Number of New Drop-Outs from MBO [persons]": "Int64",
        "Percentage of New School Drop-Outs, MBO [percent (%)]": float,
        "Number of Education Participants VAVO [persons]": "Int64",
        "Number of New Drop-Outs from VAVO [persons]": "Int64",
        "Percentage of New Drop-Outs from VAVO [percent (%)]": float,
        "Total Nuisance [number]": "Int64",
        "Youth Nuisance [number]": "Int64",
        "Nuisance Related to Alcohol/Drugs [number]": "Int64",
        "Public Drunkenness [number]": "Int64",
        "Nuisance from Homeless People [number]": "Int64",
        "Nuisance from Disturbed Persons [number]": "Int64",
        "Noise Nuisance [number]": "Int64",
        "Score for 'Safety in the Neighborhood' [score]": float,
        "Feels Unsafe in Own Neighborhood Occasionally [percent (%)]": float,
        "Feels Unsafe in Own Neighborhood Frequently [percent (%)]": float,
        "Feels Unsafe in General Occasionally [percent (%)]": float,
        "Feels Unsafe in General Frequently [percent (%)]": float
}

for column, dtype in column_data_types.items():
        try:
            data[column] =  data[column].astype(dtype)
        except Exception as e:
            print(f"Error casting column {column}: {e}")

        
data.head(5)

district_name  Year  \
0  14 Bohemen, Meer en Bos  2014   
1  14 Bohemen, Meer en Bos  2015   
2  14 Bohemen, Meer en Bos  2016   
3  14 Bohemen, Meer en Bos  2017   
4  14 Bohemen, Meer en Bos  2018   

   Average Total Score Socioeconomic Status WOA (SES/WOA Score) [score]  \
0                                              0.127                      
1                                              0.125                      
2                                              0.121                      
3                                              0.119                      
4                                              0.110                      

   Average SES/WOA Score: Financial Prosperity Subscore [score]  \
0                                              0.105              
1                                              0.104              
2                                              0.092              
3                                              0.087              
4                                              0.085              

   Average SES/WOA Score: Educational Level Subscore [score]  \
0                                              0.034           
1                                              0.035           
2                                              0.037           
3                                              0.041           
4                                              0.041           

   Average SES/WOA Score: Work History Subscore [score]  \
0                                             -0.013      
1                                             -0.013      
2                                             -0.007      
3                                             -0.009      
4                                             -0.016      

   Gross Population Density [residents/hectare]  0 to 19-Year-Olds [persons]  \
0                                          38.5                          627   
1                                          39.1                          602   
2                                          38.9                          612   
3                                          39.3                          622   
4                                          40.0                          648   

   20 to 64-Year-Olds [persons]  65-Year-Olds and Above [persons]  ...  \
0                          2408                              1637  ...   
1                          2399                              1743  ...   
2                          2373                              1773  ...   
3                          2400                              1793  ...   
4                          2415                              1829  ...   

   Nuisance Related to Alcohol/Drugs [number]  Public Drunkenness [number]  \
0                                           0                            1   
1                                           2                            1   
2                                           5                            0   
3                                          10                            2   
4                                          12                            2   

   Nuisance from Homeless People [number]  \
0                                       0   
1                                       0   
2                                       4   
3                                      10   
4                                      10   

   Nuisance from Disturbed Persons [number]  Noise Nuisance [number]  \
0                                        17                        9   
1                                        26                        7   
2                                        32                        6   
3                                        62                       19   
4                                        47                       23   

   Score for 'Safety in the Neighborhood' [score]  \
0                                             NaN 

In [3]:
data.tail(5)

district_name  Year  \
479  44 Leidschenveen  2020   
480  44 Leidschenveen  2021   
481  44 Leidschenveen  2022   
482  44 Leidschenveen  2023   
483  44 Leidschenveen  2024   

     Average Total Score Socioeconomic Status WOA (SES/WOA Score) [score]  \
479                                              0.327                      
480                                              0.318                      
481                                                NaN                      
482                                                NaN                      
483                                                NaN                      

     Average SES/WOA Score: Financial Prosperity Subscore [score]  \
479                                              0.131              
480                                              0.128              
481                                                NaN              
482                                                NaN              
483                                                NaN              

     Average SES/WOA Score: Educational Level Subscore [score]  \
479                                              0.094           
480                                              0.092           
481                                                NaN           
482                                                NaN           
483                                                NaN           

     Average SES/WOA Score: Work History Subscore [score]  \
479                                              0.103      
480                                              0.099      
481                                                NaN      
482                                                NaN      
483                                                NaN      

     Gross Population Density [residents/hectare]  \
479                                          61.8   
480                                          61.5   
481                                          61.2   
482                                          61.0   
483                                          60.5   

     0 to 19-Year-Olds [persons]  20 to 64-Year-Olds [persons]  \
479                         6402                         12543   
480                         6230                         12459   
481                         6041                         12428   
482                         5803                         12508   
483                         5631                         12388   

     65-Year-Olds and Above [persons]  ...  \
479                              1951  ...   
480                              2114  ...   
481                              2207  ...   
482                              2323  ...   
483                              2444  ...   

     Nuisance Related to Alcohol/Drugs [number]  Public Drunkenness [number]  \
479                                          16                            5   
480                                          16                            2   
481                                          29                            5   
482                                          25                            1   
483                                        <NA>                         <NA>   

     Nuisance from Homeless People [number]  \
479                                      19   
480                                       3   
481                                       7   
482                                       6   
483                                    <NA>   

     Nuisance from Disturbed Persons [number]  Noise Nuisance [number]  \
479                                        76                      124   
480                                        57                      128   
481                                        46                       56   
482                                        55                       63   
483                                      <NA>       

In [4]:
data.info()

print(f'\nThere are ',data['district_name'].nunique(),' different districts')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 484 entries, 0 to 483
Data columns (total 78 columns):
 #   Column                                                                 Non-Null Count  Dtype  
---  ------                                                                 --------------  -----  
 0   district_name                                                          484 non-null    object 
 1   Year                                                                   484 non-null    Int64  
 2   Average Total Score Socioeconomic Status WOA (SES/WOA Score) [score]   332 non-null    float64
 3   Average SES/WOA Score: Financial Prosperity Subscore [score]           332 non-null    float64
 4   Average SES/WOA Score: Educational Level Subscore [score]              332 non-null    float64
 5   Average SES/WOA Score: Work History Subscore [score]                   332 non-null    float64
 6   Gross Population Density [residents/hectare]                           476 non-null    flo

- The dataset consists of 78 columns and 484 rows, capturing various metrics at the district level.
- It combines demographic, socioeconomic, educational, and nuisance-related data.

**Demographic Data:**

- Columns include population counts across age groups (e.g., "0 to 19-Year-Olds", "65-Year-Olds and Above").
- Metrics include percentages for these age groups relative to the total population and average age of residents.

**Socioeconomic Data:**

- Includes socioeconomic status scores and subscores (e.g., financial prosperity, educational level, work history).
- Provides insights into income levels for households and individuals (e.g., disposable income, personal income for 52 weeks).

**Educational Data:**

- Tracks education levels (low, intermediate, high) and access to schools and daycare facilities (e.g., distances to schools, number of facilities within 1-10 km).
- Includes data on education participants and drop-outs, with additional details for secondary education and vocational training (MBO).

**Nuisance and Safety Data:**

- Captures reports of nuisance related to alcohol, drugs, homelessness, and disturbances.
- Includes metrics on noise complaints, public drunkenness, and safety perceptions (e.g., "Feels Unsafe in Own Neighborhood").

 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

<span style="color:red"> **Data Types:** </span>

- Columns include categorical data (e.g., "District Name"), numeric data (e.g., integers and floats), and nullable types.

<span style="color:red">**Data Completeness:**</span>

- Some columns, such as population and nuisance metrics, are well-populated, while others (e.g., job seekers, safety scores) have significant missing values, making imputation or targeted analysis necessary.

<br/>

***

# ``Task 2: Hypothesis formulation``

Think about your focus of analysis. What do you want to find out? A first step is to think about what variables might be related with each other and what these relationships could like like. Think also about a causal theory. Why do you expect this relationship to exist and to take on a specific form.

To give you an example from another dataset (this is a dataset of the worldbank: http://data.worldbank.org/data-catalog/world-development-indicators), have a look at the hypothesis below. I chose an analysis focus and then made some asumptions on the relationships between the focus variables. These asumptions are what I want to test as a formal hypothesis later on.

```text
My hypothesis
I’d like to examine world broadband access. For that reason I chose a broadband account variable. The data is organized by country. I want to control for the wealth, population, and land area of the country. I also have a hypothesis that more urban countries are more likely to have good broadband services. There are economies of scale when providing services to a large city.

I hypothesize that larger countries have lesser access, since it is expensive to provide access over larger areas. On the contrary, countries with a lot of urban land area can take advantage of economies of scale resulting in relatively more broadband users concentrated in smaller zones within the country. We also hypothesize that wealthier countries have better broadband access, since there is a larger market to provide the newest services. A final variable which we add is rail lines. I hypothesize that broadband lines can take advantage of existing infrastructure right-of-ways, of which rail is a surrogate measure. Furthermore, the presence of rail lines may indicate other factors including a geography which is conclusive to physical development, and favourable institutional factors which promote high technology development.

My choice of variables were,

| Variable Name                 | Variable Code     |
| ----------------------------- | ----------------- |
| Fixed broadband subscriptions | IT.NET.BBND       |
| GDP (current US$)             | NY.GDP.MKTP.CD    |
| Population, total             | SP.POP.TOTL       |
| Land area (sq. km)            | AG.LND.TOTL.K2    |
| Urban land area (sq. km)      | AG.LND.TOTL.UR.K2 |
| Rail lines (total route-km)   | IS.RRS.TOTL.KM    |

```

Formulate a working hypothesis. Provide a list of variables you need to include in your hypothesis. Explain what your dependent variable is. Explain your choice of independent and control variables. Would you wish to include any variables that is not available? Are there any variables (proxies) present with which you could replace these missing values? Would that have an effect on your outcomes? Think about bias in the data. Why do you expect to find the relationship that you are hypothesising? Can you think of any cases in which this hypothesis will be rejected? Why? Can you think of any cases in which the hypothesis will be falsely rejected or falsely acepted? Think about bias in the data and in your own reasoning.

## ``Exercise: Hypothesising``

- State your hypothesis in a markdown cell as I showed in the example above (there is no single right hypothesis, you are free to make a **reasonable** choice for this task)
- Find the variables of interest for your hypothesis and mention them in the markdown cell (4-7 variables)
- Explain what your dependent variable is. Explain your choice of independent and control variables.
- Would you wish to include any variables that is not available? Are there any variables (proxies) present with which you could replace these missing values? Would that have an effect on your outcomes? Think about bias in the data. Explain your reasoning.
- Can you think of any cases in which this hypothesis will be rejected? Why?
- Can you think of any cases in which the hypothesis will be falsely rejected or falsely acepted? Think about bias in the data and in your own reasoning.
- Can you shortly reflect if there are any potentially important perspectives that you are not taking into account by choosing this hypothesis?

## The Impact of Population Density, Schools, and Crime on Dropout Rates Over Time

I hypothesize that districts with higher population density and fewer schools per capita have higher dropout rates over time. Overcrowded schools and limited resources make it harder for students to stay in school. I also believe that higher crime rates contribute to dropout rates by creating unsafe environments that discourage regular attendance and engagement.
Moreover, I think that the presence of out-of-school centers may help reduce dropout rates, especially in areas with fewer schools, by providing alternative spaces for learning. I also expect that districts with higher median household incomes and more educated adults have fewer dropouts because families in these areas are better equipped to support education.
Over multiple years, I expect to see changes in dropout rates correlate with shifts in these factors, such as the opening of new schools, improvements in public safety, or changes in economic conditions.

| **Variable Name**                     | **Dataset Variable**                                        | **Description**                                                                                 |
|:--------------------------------------|:-----------------------------------------------------------|:------------------------------------------------------------------------------------------------|
| **"school_dropout_rate"**               | Percentage of New School Drop-Outs [percent (%)]           | <span style="color:blue">Dependent</span>: Measuring the percentage of students leaving school before completion.       |
| **student_population**                | Number of 0 to 23-Year-Olds in The Hague Population [persons] | <span style="color:green">Independent</span>: Reflecting the concentration of people per square kilometer.               |
| **schools_within_3km**     | Number of Schools Within 3 km [number]                     | <span style="color:green">Independent</span>: Indicating the availability of schools relative to the student population.         |
| **out_of_school_centers**   | Number of Out-of-School Care Centres Within 3 km [number]   | <span style="color:green">Independent</span>: Capturing alternative learning spaces that could reduce dropout rates.      |
| **crime_rate**                        | Youth Nuisance [number]                                    | <span style="color:green">Independent</span>: Measuring local safety, which may influence school attendance.             |
| **median_household_income**           | Average Disposable Income of Private Households [euros]    | <span style="color:orange">Control</span>: Accounting for socioeconomic differences between districts.                    |
| **education_attainment_level** | Average SES/WOA Score: Educational Level Subscore [score] | <span style="color:orange">Control</span>: Reflecting adult education levels, which can influence the value placed on education.|




---
**Missing or Additional Variables**

**Desired Variables**:

  - School funding or resources per student: A more direct measure of the impact of overcrowding or resource scarcity.
  - Teacher-to-student ratio: Could provide a better proxy for school overcrowding.
  
**Proxies in the Dataset**:

  - The number of schools per capita can partially act as a proxy for resource availability, but it may not fully capture the quality of education or resource disparities.
  - Crime rate serves as a proxy for safety but may not reflect other factors affecting attendance, such as bullying or lack of transportation.

---

**Potential Rejections of the Hypothesis**

In this case the hypothesis may be validly rejected if the data shows no significant correlation between factors like population density, school availability, crime rates, and dropout rates. For instance, if dropout rates remain low in densely populated areas with fewer schools, it would contradict the hypothesis. However, it could be falsely rejected due to biases in data collection, such as underreporting dropout rates in certain districts, or the use of proxies like crime rate, which may not fully capture other safety issues like bullying or unsafe commutes. Similarly, the hypothesis could be falsely accepted if unrelated interventions, such as government subsidies or national campaigns promoting education, reduce dropout rates independently of the analyzed factors.

---

**Potentially Missing Perspectives**

- Qualitative aspects: Factors like cultural values, parental involvement, and teacher motivation are not captured.
- Temporal effects: Delayed impacts of interventions (e.g., opening new schools) may not align with the study's timeframe.
- Equity considerations: Differences in dropout rates among socioeconomic or ethnic groups may provide additional insights into systemic biases or inequities.








<br/>

***

# ``Task 2: Subsetting the Data``




From now on we want to work with a subset of the data that is relevant to your study focus. Choose **4 to 7 variables** from your dataset for further exploratory statistics (more information will follow later in exercises). Think carefully about how you want to address the neighboorhoods. Does it make sense to build a new ID or should you use the provided IDs? Remember that it is important to work with codes as opposed to names to make our analyses more reproducible across time. For now let’s set aside the added complexity of time series and dynamics. Our task is to select just one year.


## ``Exercise: Subsetting the Data``

- Subset your dataframe based on your focus of analysis and your choice of variables.
- your dataframe would have greatly reduced in size and looks neater, show us what it looks like now using head() or something similar
    - show some statistics like number or rows, columns, names of variables and unique neighbourhoods, etc.

You can count the columns manually, but in a large data set like this it is accurate and convenient to let python calculate this for us. Get the index of relevant columns and store them in a variable.




In [5]:
# List of columns to extract
columns_to_extract = [
    "district_name",
    "Year",
    "Percentage of New School Drop-Outs [percent (%)]",
    "Number of 0 to 23-Year-Olds in The Hague Population [persons]",
    "Number of Schools Within 3 km [number]",
    "Number of Out-of-School Care Centres Within 3 km [number]",
    "Youth Nuisance [number]",
    "Average Disposable Income of Private Households [euros]",
    "Average SES/WOA Score: Educational Level Subscore [score]"
]

# Extract the required columns
filtered_data = data[columns_to_extract]
filtered_data.head(10)


district_name  Year  \
0  14 Bohemen, Meer en Bos  2014   
1  14 Bohemen, Meer en Bos  2015   
2  14 Bohemen, Meer en Bos  2016   
3  14 Bohemen, Meer en Bos  2017   
4  14 Bohemen, Meer en Bos  2018   
5  14 Bohemen, Meer en Bos  2019   
6  14 Bohemen, Meer en Bos  2020   
7  14 Bohemen, Meer en Bos  2021   
8  14 Bohemen, Meer en Bos  2022   
9  14 Bohemen, Meer en Bos  2023   

   Percentage of New School Drop-Outs [percent (%)]  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   
5                                             0.025   
6                                               NaN   
7                                             0.025   
8                                             0.026   
9                                               NaN   

   Number of 0 to 23-Year-Olds in The Hague Population [persons]  \
0                                               <NA>               
1                                                688               
2                                                696               
3                                                717               
4                                                730               
5                                                684               
6                                                703               
7                                                681               
8                                                710               
9                                                697               

   Number of Schools Within 3 km [number]  \
0                                    <NA>   
1                                    <NA>   
2                                    <NA>   
3                                    <NA>   
4                                    <NA>   
5                                      24   
6                                      24   
7                                      24   
8                                      20   
9                                    <NA>   

   Number of Out-of-School Care Centres Within 3 km [number]  \
0                                               <NA>           
1                                               <NA>           
2                                               <NA>           
3                                               <NA>           
4                                               <NA>           
5                                                 39           
6                                                 37           
7                                                 39           
8                                                 34           
9                                               <NA>           

   Youth Nuisance [number]  \
0                        2   
1                        2   
2                        6   
3                       18   
4                       18   
5                       15   
6                       50   
7                       40   
8                       21   
9                       32   

   Average Disposable Income of Private Households [euros]  \
0                                            38400.0         
1                                            38500.0         
2                                            39100.0         
3                                            40400.0         
4                                            41300.0         
5                                            43300.0         
6                                            44800.0         
7                                            46000.0         
8                                                NaN         
9                                                NaN         

   Average SES/WOA Score: Educational Level Su

<br/>

***

# ``Task 3: Manage Missing Data``

Datasets often have missing data. Not all variables are as well documented for every year as others. There are a number of ways in general of dealing with missing data. These involve

1. Dropping off cases (or rows) in the data with any missing variables
2. Excluding variables in the data with any missing data
3. Selectively choosing indicators with only a limited amount of missing data
4. Replacing missing variables with averages, or other representative values
5. Creating a separate model to predict missing data

In this assignment we are going to use a number of these strategies. We can certainly be dropping off cases (strategy one). I am loathe to drop off whole indicators. But we can, for example, choose a year for the indicator where most of the data is available (strategy three).

Building a separate model to impute missing data, is often a good idea. But that requires a first working model before we even consider building a missing data model (and we haven't gotten there yet in this course); the working model and the missing data model are often constructed together. Note also that there are packages in Python which will construct a model of your data, and then impute missing values for you. You may or may not find these functions and packaging for modelling your data to be fully appropriate. Therefore treat these missing data models very seriously, and not as a black box. Models of missing data are as important, and deserve just as much care and caution as any other statistical model.

In the next section I discuss some specifics about how the data is currently formatted, and how we would like to have it formatted for analysis purposes.

## ``Exercise: Manage Missing Data``
- Explore and handle the missing data in your data set using one of the methods mentioned above. Visualize your findings if necessary. Think about potential explanations for gaps in the data. Think about what stage of the process led to the data gaps? Are the gaps going to be a problem for your analysis? Why?
- Explain your choice of method. Can you think of ways how the chosen method skews your results? Did your data quality increase? Why?
- At this stage, reflect on your choices along the way. Should you reiterate (e.g. choose a different focus/ hypothesis/ year for your analysis)? Why?
- Also shortly reflect what the data gaps might mean. Are there collections of variables for which data is structurally missing? Are there neighbourhoods for which data is structurally missing? What can cause these structural gaps in the data?

-----
I decided to focus on the year 2019 because it had the most complete data. To ensure the analysis remains accurate, I dropped rows where important independent variables, like student population or number of schools, were missing since these directly affect the results. For other missing values, like dropout rates or household income, I used mean imputation to fill the gaps, keeping as much data as possible while maintaining consistency. Some data gaps may be due to issues like underreporting in certain districts. I will monitor potential bias from these methods and make adjustments if needed.

In [6]:
# Step 1: Group by 'Year' and count non-null values for each column
non_null_counts_by_year = filtered_data.groupby("Year").apply(lambda x: x.notnull().sum(axis=0))

# Step 2: Calculate the total number of non-null values across all columns for each year
total_non_null_by_year = non_null_counts_by_year.sum(axis=1)

# Step 3: Find the year with the most available data
year_with_most_data = total_non_null_by_year.idxmax()
max_value = total_non_null_by_year.max()

# Output the result
print(f"The year with the most available data is {year_with_most_data} with {max_value} non-null values.")

The year with the most available data is 2019 with 375 non-null values.


In [7]:
# Step 4: Filter the dataset for the year with the most available data
filtered_data_for_best_year = filtered_data[filtered_data["Year"] == 2019]

# Step 5: Drop the 'Year' column
filtered_data_for_best_year = filtered_data_for_best_year.drop(columns=["Year"])

filtered_data_for_best_year.head()

filtered_data_for_best_year.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44 entries, 5 to 478
Data columns (total 8 columns):
 #   Column                                                         Non-Null Count  Dtype  
---  ------                                                         --------------  -----  
 0   district_name                                                  44 non-null     object 
 1   Percentage of New School Drop-Outs [percent (%)]               31 non-null     float64
 2   Number of 0 to 23-Year-Olds in The Hague Population [persons]  42 non-null     Int64  
 3   Number of Schools Within 3 km [number]                         43 non-null     Int64  
 4   Number of Out-of-School Care Centres Within 3 km [number]      43 non-null     Int64  
 5   Youth Nuisance [number]                                        44 non-null     Int64  
 6   Average Disposable Income of Private Households [euros]        42 non-null     float64
 7   Average SES/WOA Score: Educational Level Subscore [score]      4

In [8]:
#Drop rows where independent variables are missing and fill where
# Step 1: Drop rows where independent variables are missing
independent_vars = [
    "Number of 0 to 23-Year-Olds in The Hague Population [persons]",
    "Number of Schools Within 3 km [number]",
    "Number of Out-of-School Care Centres Within 3 km [number]",
    "Youth Nuisance [number]"
]
# Drop rows with missing values in independent variables
cleaned_dataset = filtered_data_for_best_year.dropna(subset=independent_vars).reset_index(drop=True)  # Fully reset index for the new DataFrame


# Step 2: Fill missing values for other columns (mean imputation for dependent/control variables)
variables_to_impute = [
    "Percentage of New School Drop-Outs [percent (%)]",
    "Average Disposable Income of Private Households [euros]",
    "Average SES/WOA Score: Educational Level Subscore [score]"
]
for col in variables_to_impute:
    cleaned_dataset.loc[:, col] = cleaned_dataset[col].fillna(cleaned_dataset[col].mean())


## ``Task 4: Reshape the Data``
Reshaping data is a two-step process of melting and pivoting the data. Melting the data involves describing which data are indicators ("id") and which are variables for retrieval (“measure”). In this case your data may already be in melted form (long form). Pivoting then involves actually reshaping the data into the needed format. In this step, you have to reshape the data from long to wide format.

Pivoting the data involves specifying what data is on the rows and on the columns. Hint: functions melt and pivot offered by ``numpy`` library in python. For our analyses we want the neighborhood IDs to be on the rows, and to have all 4-7 variables as columns, where the value of each cell is the value taken from the column year that you chose at the subsetting step.

## ``Exercise: Reshape the Data``
- Do you need to melt and pivot the data in your specific case? Explain.
- Examine the dimensions of the new datadrame if you had to pivot. Show it to us using head or print commands.
- Then rename all column names to something better and useful and easily addressable.

In [9]:
final_dataset = cleaned_dataset

#setting the district names to be use as id
final_dataset = final_dataset.set_index('district_name')


#rename the columns 
final_dataset.rename(columns={
    "Percentage of New School Drop-Outs [percent (%)]": "school_dropout_rate",
    "Number of 0 to 23-Year-Olds in The Hague Population [persons]": "student_population",
    "Number of Schools Within 3 km [number]": "schools_within_3km",
    "Number of Out-of-School Care Centres Within 3 km [number]": "out_of_school_centers",
    "Youth Nuisance [number]": "crime_rate",
    "Average Disposable Income of Private Households [euros]": "median_household_income",
    "Average SES/WOA Score: Educational Level Subscore [score]": "education_attainment_level"
}, inplace=True)

final_dataset.head(10)

school_dropout_rate  student_population  \
district_name                                                           
14 Bohemen, Meer en Bos                  0.025000                 684   
15 Kijkduin en Ockenburgh                0.022742                 489   
16 Kraayenstein en Vroondaal             0.020000                1560   
17 Loosduinen                            0.029000                4164   
18 Waldeck                               0.023000                3227   
31 Rustenburg en Oostbroek               0.024000                5094   
32 Leyenburg                             0.023000                3376   
33 Bouwlust/Vrederust                    0.024000                8672   
34 Morgenstond                           0.020000                5198   
36 Moerwijk                              0.032000                6348   

                              schools_within_3km  out_of_school_centers  \
district_name                                                             
14 Bohemen, Meer en Bos                       24                     39   
15 Kijkduin en Ockenburgh                     14                     25   
16 Kraayenstein en Vroondaal                  11                     15   
17 Loosduinen                                 27                     38   
18 Waldeck                                    25                     39   
31 Rustenburg en Oostbroek                    60                     84   
32 Leyenburg                                  54                     72   
33 Bouwlust/Vrederust                         22                     33   
34 Morgenstond                                37                     53   
36 Moerwijk                                   45                     60   

                              crime_rate  median_household_income  \
district_name                                                       
14 Bohemen, Meer en Bos               15                  43300.0   
15 Kijkduin en Ockenburgh              7                  77200.0   
16 Kraayenstein en Vroondaal          18                  57800.0   
17 Loosduinen                         85                  40700.0   
18 Waldeck                            76                  36400.0   
31 Rustenburg en Oostbroek            57                  35000.0   
32 Leyenburg                         121                  37600.0   
33 Bouwlust/Vrederust                167                  33100.0   
34 Morgenstond                        62                  32000.0   
36 Moerwijk                           70                  28800.0   

                              education_attainment_level  
district_name                                             
14 Bohemen, Meer en Bos                            0.048  
15 Kijkduin en Ockenburgh                          0.080  
16 Kraayenstein en Vroondaal                       0.064  
17 Loosduinen                                     -0.038  
18 Waldeck                                        -0.045  
31 Rustenburg en Oostbroek                        -0.065  
32 Leyenburg                                      -0.029  
33 Bouwlust/Vrederust                             -0.156  
34 Morgenstond                                    -0.141  
36 Moerwijk                                       -0.182

<br/>

***

# ``Task 5: Saving the Results``

_Note:_ We do not need this file but we expect that if you learn how to save your data, it will be very useful in the future, as you do not need to run the script to clean your data again.

- Save the cleaned dataframe as 'assignment-01-cleaned.csv' in your data folder
- Consider your final dataset. Reflect on why you think that this newly created dataset is appropriate/ not appropriate to asnwer your research question.


In [10]:
final_dataset.to_csv('data/assignment-01-cleaned.csv', index = True)